bauhaus/utils.py

```diff
@@ -141,6 +143,14 @@ def unpack_variables(T, propositions) -> list:
             inputs.update(unpack_variables(var, propositions))
 
         else:
+            # consider CustomNNFs that only store a single Var
+            if isinstance(var, core.CustomNNF):
+                if len(var.args) == 1:
```

Comment:
What's this checking? That it's a literal? I'm worried that an `And` or an `Or` with just a single argument may cause problems. A `Not` will always have one argument too. Maybe appeal to the `var.typ` value? See [here](https://github.com/QuMuLab/bauhaus/blob/main/bauhaus/core.py#L285).

Reply:
@beckydvn It a decent time to *nudge* before the semester begins? ;)

Happy new years, btw!

Reply:
Whoops sorry, totally forgot about this!

Honestly it's been quite a while so I forget haha, I'll take a look now! :)

Reply:
@haz Hmm... After taking another look and reassessing the issue, I think it's best to get your input before writing anything that may be unnecessary. Sorry this is a little long, I'll try to walk you through my thought process.

What I was doing initially was checking if a `CustomNNF`was nesting something that could be turned into a `Var` or its negation. This includes `Ands` and `Ors` that are only holding a single proposition, as well as `Nots` that are applied to a single proposition. I was motivated to make this PR to solve the issue where the following can happen (side note: I vaguely remember that when I was trying to solve SLAF using bauhaus, I was ending up with thousands of duplicate variables and solving became unfeasible. I believe this may be why).

```
g = Encoding()
@proposition(g)
class H:
    def __init__(self, data):
        self.data = data

constraint.add_exactly_one(g, H(1))
constraint.add_exactly_one(g, ~H(1))
g.pprint(g.compile()) # prints (Var(1) ∧ (Var(<bauhaus.core.CustomNNF object at 0x7f098daaf040>)))
assert g.compile().solve() == None # fails
```

Since `~H(1)` is wrapped in a `CustomNNF`, the solver doesn't recognize that they are holding the same variable and finds a solution. My solution fixes this code by converting to `Vars` where possible, returning the compiled formula `(Var(1) ∧ Var(¬1))` and not finding a solution. Obviously the code I wrote so far only works in the case where `CustomNNF` is nesting a single variable, and would be done recursively/on each argument to "unpack" the `CustomNNFs` as much as possible.

That's when I noticed that the issue only happens with "bauhaus" constraints and never "custom" ones, added like `g.add_constraint(H(1) & ~H(1))`, as when they compile they call this compile function: https://github.com/QuMuLab/bauhaus/blob/ad56897d81034dcff8366628db554cb29946eb5e/bauhaus/core.py#L319
which I now realize is the function I was going to write 😁

My question is, is there any reason why "bauhaus" constraints don't ever call this function when compiling and have to stay in their `CustomNNF` form? I believe it would be a simple fix to the issues above, but perhaps there are other reasons, design/otherwise why this wasn't done initially. If they _must_ stay in `CustomNNF` form then I suppose the above isn't an actual issue and this PR can be closed.

Let me know what you think, and apologies for the long read!

Reply:
Hrmz...I think they should always be compiled before the testing should check out. What might be happening in the example above is that equality is done on the object creation level. What if you changed the proposition to be...

```python
@proposition(g)
class H:
    def __init__(self, data):
        self.data = data
    def __eq__(self, other):
        return isinstance(other, self.__class__) and hash(self) == hash(other)
    def __hash__(self):
        return hash(self.data)
```

Reply:
I actually had those functions in the class already, just deleted them in the code segment above for brevity 😅

Reply:
Aha!!

```python
from bauhaus import constraint, proposition, Encoding

E = Encoding()

@proposition(E)
class H:
    def __init__(self, data):
        self.data = data
    def __eq__(self, other):
        return isinstance(other, self.__class__) and hash(self) == hash(other)
    def __hash__(self):
        return hash(self.data)
    def __repr__(self) -> str:
        return str(self.data)
    def __str__(self) -> str:
        return str(self.data)

# E.add_constraint(H(1))
# E.add_constraint(~H(1))

constraint.add_exactly_one(E, H(1))
constraint.add_exactly_one(E, ~H(1))

constraint.add_exactly_one(E, H(2), 123)

E.pprint(E.compile())

E.introspect()
```

So what's happening here is that _anything_ you put in the `exactly_one` is being treated as a propositional variable. Doesn't exist yet? Create a new propositional variable from the object. So the `~H(1)` you are using is seen as an arbitrary object that should represent its own variable.

Technically the `add_exactly_one` stipulates that you should only provide variables. Maybe we could confirm any argument to a regular constraint _isn't_ a `CustomNNF`? Can't think of a reason for the existing `constraint.add_*` methods to use literals (i.e., allowing negation) instead of variables.

Reply:
Makes sense! I thought they could take both literals and variables which is where my confusion arose...

So, raise an exception if they provide a `CustomNNF`? Or maybe just a warning explaining their mistake and give them the `Var(<memory address>)` by default?

Reply:
Maybe a warning when an encoding is compiled. You could check the python-nnf variables for anything of CustomNNF class.